### ISSUE TITLE
CNN training is much slower than PyTorch on the same MPS backend

### ISSUE THREAD

**Issue by JacobCote** [tracel-ai/burn#4875]
## Summary

I recently had performance issues with a vision model so I created a somewhat small minimal repro setup. I benchmarked a small CNN training loop on Apple MPS using equivalent Burn and PyTorch repros. Even when both use the same backend family (burn-tch / libtorch vs PyTorch on MPS) and similar model, loss, optimizer, and tensor shapes, the Burn version is significantly slower.

## Results

Average results after a 20 step warmup: 

- Burn: `514.61 ms/step`
- PyTorch: `165.40 ms/step`

## Environment

- Apple M2
- 16 GB RAM
- macOS 15.3.1
- PyTorch `2.9.0`
- Burn `0.21.0-pre.2`

## Burn repro

Create a new Cargo project with this `Cargo.toml`:

```toml
[package]
name = "cnn_min_repro"
version = "0.1.0"
edition = "2024"

[dependencies]
burn = { version = "0.21.0-pre.2" }

[features]
default = []
tch-gpu = ["burn/tch", "burn/train"]
```

Put this code in `src/main.rs` and run with:

```bash
cargo run --release --no-default-features --features tch-gpu
```

```rust
#![recursion_limit = "256"]

use std::time::Instant;

use burn::backend::Autodiff;
use burn::backend::libtorch::{LibTorch, LibTorchDevice};
use burn::module::Module;
use burn::nn::conv::{Conv2d, Conv2dConfig};
use burn::nn::loss::{MseLoss, Reduction};
use burn::nn::pool::{MaxPool2d as Pool2d, MaxPool2dConfig as Pool2dConfig};
use burn::nn::PaddingConfig2d;
use burn::optim::{AdamConfig, GradientsParams, Optimizer};
use burn::tensor::activation::{relu, sigmoid};
use burn::tensor::backend::Backend;
use burn::tensor::Tensor;

type MyBackend = Autodiff<LibTorch<f32>>;

const BATCH_SIZE: usize = 32;
const INPUT_SIZE: usize = 256;
const GRID_SIZE: usize = 16;
const LEARNING_RATE: f64 = 3e-4;
const WARMUP_STEPS: usize = 20;
const MEASURE_STEPS: usize = 40;

#[derive(Debug, Module)]
struct ConvNet<B: Backend> {
    conv1: Conv2d<B>,
    conv2: Conv2d<B>,
    conv3: Conv2d<B>,
    conv4: Conv2d<B>,
    head: Conv2d<B>,
    pool: Pool2d,
}

impl<B: Backend> ConvNet<B> {
    fn new(device: &B::Device) -> Self {
        let conv1 = Conv2dConfig::new([3, 32], [3, 3])
            .with_padding(PaddingConfig2d::Same)
            .init(device);
        let conv2 = Conv2dConfig::new([32, 64], [3, 3])
            .with_padding(PaddingConfig2d::Same)
            .init(device);
        let conv3 = Conv2dConfig::new([64, 128], [3, 3])
            .with_padding(PaddingConfig2d::Same)
            .init(device);
        let conv4 = Conv2dConfig::new([128, 128], [3, 3])
            .with_padding(PaddingConfig2d::Same)
            .init(device);
        let head = Conv2dConfig::new([128, 1], [1, 1]).init(device);
        let pool = Pool2dConfig::new([2, 2]).init();

        Self {
            conv1,
            conv2,
            conv3,
            conv4,
            head,
            pool,
        }
    }

    fn forward(&self, input: Tensor<B, 4>) -> Tensor<B, 3> {
        let x = self.pool.forward(relu(self.conv1.forward(input)));
        let x = self.pool.forward(relu(self.conv2.forward(x)));
        let x = self.pool.forward(relu(self.conv3.forward(x)));
        let x = self.pool.forward(relu(self.conv4.forward(x)));
        let x = sigmoid(self.head.forward(x));
        x.squeeze_dim::<3>(1)
    }
}

fn main() {
    let device = LibTorchDevice::Mps;
    let mut model = ConvNet::<MyBackend>::new(&device);
    let mut optimizer = AdamConfig::new().init();
    let criterion = MseLoss::new();
    let input = Tensor::<MyBackend, 4>::zeros([BATCH_SIZE, 3, INPUT_SIZE, INPUT_SIZE], &device);
    let target = Tensor::<MyBackend, 3>::zeros([BATCH_SIZE, GRID_SIZE, GRID_SIZE], &device);

    println!("Using device: {:?}", device);
    println!("Trainable parameters: {}", model.num_params());

    for _ in 0..WARMUP_STEPS {
        let preds = model.forward(input.clone());
        let loss = criterion.forward(preds, target.clone(), Reduction::Mean);
        let grads = loss.backward();
        let grads = GradientsParams::from_grads(grads, &model);
        model = optimizer.step(LEARNING_RATE, model, grads);
    }

    let start = Instant::now();
    for _ in 0..MEASURE_STEPS {
        let preds = model.forward(input.clone());
        let loss = criterion.forward(preds, target.clone(), Reduction::Mean);
        let grads = loss.backward();
        let grads = GradientsParams::from_grads(grads, &model);
        model = optimizer.step(LEARNING_RATE, model, grads);
    }

    let total_ms = start.elapsed().as_secs_f64() * 1000.0;
    println!("Measured {} training steps, avg: {:.2} ms", MEASURE_STEPS, total_ms / MEASURE_STEPS as f64);
}
```

## PyTorch repro

Put this code in `min_repro.py` and run with:

```bash
python3 min_repro.py mps
```

```python
from __future__ import annotations

import sys
import time
import torch
import torch.nn as nn

BATCH_SIZE = 32
INPUT_SIZE = 256
GRID_SIZE = 16
WARMUP_STEPS = 20
MEASURE_STEPS = 40
LEARNING_RATE = 3e-4
DEVICE = sys.argv[1]


class ConvNet(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 32, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.conv3 = nn.Conv2d(64, 128, kernel_size=3, padding=1)
        self.conv4 = nn.Conv2d(128, 128, kernel_size=3, padding=1)
        self.head = nn.Conv2d(128, 1, kernel_size=1)
        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
        self.act = nn.ReLU(inplace=True)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.pool(self.act(self.conv1(x)))
        x = self.pool(self.act(self.conv2(x)))
        x = self.pool(self.act(self.conv3(x)))
        x = self.pool(self.act(self.conv4(x)))
        x = torch.sigmoid(self.head(x))
        return x.squeeze(1)


def count_trainable_parameters(model: nn.Module) -> int:
    return sum(p.numel() for p in model.parameters() if p.requires_grad)


def main() -> None:
    device = torch.device(DEVICE)
    model = ConvNet().to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=LEARNING_RATE)
    criterion = nn.MSELoss()

    inputs = torch.zeros(BATCH_SIZE, 3, INPUT_SIZE, INPUT_SIZE, device=device)
    targets = torch.zeros(BATCH_SIZE, GRID_SIZE, GRID_SIZE, device=device)

    print(f"Using device: {device}")
    print(f"Trainable parameters: {count_trainable_parameters(model)}")

    model.train()
    for _ in range(WARMUP_STEPS):
        optimizer.zero_grad(set_to_none=True)
        preds = model(inputs)
        loss = criterion(preds, targets)
        loss.backward()
        optimizer.step()

    start = time.perf_counter()
    for _ in range(MEASURE_STEPS):
        optimizer.zero_grad(set_to_none=True)
        preds = model(inputs)
        loss = criterion(preds, targets)
        loss.backward()
        optimizer.step()

    total_ms = (time.perf_counter() - start) * 1000.0
    print(f"Measured {MEASURE_STEPS} training steps avg: {total_ms / MEASURE_STEPS} ms")


if __name__ == "__main__":
    main()
```
